HM4/decorators.py

```diff
@@ -0,0 +1,145 @@
+from functools import wraps
+import time
+from collections import Hashable
+
+# === Task 1 ===
+"""
+Write a decorator which wraps functions to log function arguments and the return value on each call. Provide support for
+both positional and named arguments (your wrapper function should take both *args and **kwargs and print them both):
+
+#>>> func(4, 4, 4)
+you called func(4, 4, 4)
+it returned 6 6
+"""
+print """     === Task 1 ===      \n"""  # to separate results in console
+
+
+def log(func):
+
+    @wraps(func)
+    def log_wrapper(*args, **kwargs):
+        if kwargs:
+            return 'you called: {}({}) \nit returned: {}'.format(func.__name__, str(kwargs)[1:-1], func(**kwargs))
+        else:
+            return 'you called: {} \nit returned: {}'.format(func.__name__ + str(args), func(*args))
+
+    return log_wrapper
+
+
+@log
+def multiply(x, y):
+    return x*y
+
+print multiply(2, 3)
+print multiply(y=5, x=2), '\n'*3
+
+# === Task 2 ===
+"""
+
+Write a decorator to memoize functions with an arbitrary set of arguments.
+Memoization is only possible if the arguments are hashable. Example function - Fibonacci.
+
+Memoization - if function was already called with this arguments, then get result from cache. if not - execute function,
+ store result in cache, and return result.
+
+If the wrapper is called with arguments which are not hashable, then the wrapped function should just be called without
+caching.
+
+Note: To use args and kwargs as dictionary keys, they must be hashable, which basically means that they must be
+immutable. args is already a tuple, which is fine, but kwargs have to be converted.
+One way is tuple(sorted(kwargs.items())).
+
+Such functions could be used to reduce cost of computation of some functions.
+"""
+print """     === Task 2 ===      \n"""  # to separate results in console
+
+
+def memo(func):   # not sure about proper work of this wrapper
+
+    func_memo = {}
+
+    # @wraps(func)
```

Comment:
You can use wraps without any problems. I guess you commented it because you have some Attribute error while running your unit test. Solution here - just add attribute `__name__` to your mocked function inside test(I will add  comment there also to be clear)
If you are interested why such error happens:

So when you use decorator without 'wraps' your decorated function lose original function signature(e.g. __doc__, __name__ etc.). When you use wraps - it simply take all those attrs from original function and assign those to a wrapper. Code from functools library:
```
WRAPPER_ASSIGNMENTS = ('__module__', '__name__', '__doc__')
WRAPPER_UPDATES = ('__dict__',)
def update_wrapper(wrapper,
                   wrapped,
                   assigned = WRAPPER_ASSIGNMENTS,
                   updated = WRAPPER_UPDATES):
    """Update a wrapper function to look like the wrapped function

       wrapper is the function to be updated
       wrapped is the original function
       assigned is a tuple naming the attributes assigned directly
       from the wrapped function to the wrapper function (defaults to
       functools.WRAPPER_ASSIGNMENTS)
       updated is a tuple naming the attributes of the wrapper that
       are updated with the corresponding attribute from the wrapped
       function (defaults to functools.WRAPPER_UPDATES)
    """
    for attr in assigned:
        setattr(wrapper, attr, getattr(wrapped, attr))
    for attr in updated:
        getattr(wrapper, attr).update(getattr(wrapped, attr, {}))
```
As you see, getattr method tries to extract attributes from predefined list of values, including `__name__`, that your Mocked function does not have. And if you assign it to your magic mocked function everything will work. You can ask - but I also should assign `__doc__` there? No. Mock class already have it. 

I hope this make situation more clear. 
P.S: If something looks complicated - try to check sources :)